### PR TITLE
[nexx] Download using correct protection tokens (fixes #24309)

### LINKED ISSUE #24309: Download from funk.net fails with HTTP Error 403: Forbidden / HTTP Error 400: Bad Request
## Checklist

- [x] I'm reporting a broken site support
- [x] I've verified that I'm running youtube-dl version **2020.03.08**
- [x] I've checked that all provided URLs are alive and playable in a browser
- [x] I've checked that all URLs and arguments with special characters are properly quoted or escaped
- [x] I've searched the bugtracker for similar issues including closed ones

## Verbose log

```
$ youtube-dl --verbose  https://www.funk.net/channel/doctor-who-1164/journeys-end-ov-515545/doctor-who-staffel-4-ov-2473

[debug] System config: []
[debug] User config: []
[debug] Custom config: []
[debug] Command-line args: [u'--verbose', u'https://www.funk.net/channel/doctor-who-1164/journeys-end-ov-515545/doctor-who-staffel-4-ov-2473']
[debug] Encodings: locale UTF-8, fs UTF-8, out UTF-8, pref UTF-8
[debug] youtube-dl version 2020.03.08
[debug] Python version 2.7.12 (CPython) - Linux-4.15.0-88-generic-x86_64-with-LinuxMint-18.1-serena
[debug] exe versions: ffmpeg 2.8.15, ffprobe 2.8.15, rtmpdump 2.4
[debug] Proxy map: {}
[Funk] 515545: Downloading JSON metadata
[Nexx] 515545: Downloading JSON metadata
[Nexx] 515545: Downloading m3u8 information
WARNING: Failed to download m3u8 information: HTTP Error 403: Forbidden
[Nexx] 515545: Downloading MPD manifest
WARNING: Failed to download MPD manifest: HTTP Error 403: Forbidden
[Nexx] 515545: Downloading ISM manifest
WARNING: Failed to download ISM manifest: HTTP Error 403: Forbidden
[debug] Default format spec: bestvideo+bestaudio/best
[debug] Invoking downloader on u'http://nx-d09.akamaized.net/06be6b49-c771-4f80-9ceb-a658ebc5e751/515545_src_1920x1080_6000.mp4'
ERROR: unable to download video data: HTTP Error 400: Bad Request
Traceback (most recent call last):
  File "/home/yee/Develop/youtube-dl/youtube-dl/youtube_dl/YoutubeDL.py", line 1926, in process_info
    success = dl(filename, info_dict)
  File "/home/yee/Develop/youtube-dl/youtube-dl/youtube_dl/YoutubeDL.py", line 1865, in dl
    return fd.download(name, info)
  File "/home/yee/Develop/youtube-dl/youtube-dl/youtube_dl/downloader/common.py", line 366, in download
    return self.real_download(filename, info_dict)
  File "/home/yee/Develop/youtube-dl/youtube-dl/youtube_dl/downloader/http.py", line 341, in real_download
    establish_connection()
  File "/home/yee/Develop/youtube-dl/youtube-dl/youtube_dl/downloader/http.py", line 109, in establish_connection
    ctx.data = self.ydl.urlopen(request)
  File "/home/yee/Develop/youtube-dl/youtube-dl/youtube_dl/YoutubeDL.py", line 2238, in urlopen
    return self._opener.open(req, timeout=self._socket_timeout)
  File "/usr/lib/python2.7/urllib2.py", line 435, in open
    response = meth(req, response)
  File "/usr/lib/python2.7/urllib2.py", line 548, in http_response
    'http', request, response, code, msg, hdrs)
  File "/usr/lib/python2.7/urllib2.py", line 473, in error
    return self._call_chain(*args)
  File "/usr/lib/python2.7/urllib2.py", line 407, in _call_chain
    result = func(*args)
  File "/usr/lib/python2.7/urllib2.py", line 556, in http_error_default
    raise HTTPError(req.get_full_url(), code, msg, hdrs, fp)
HTTPError: HTTP Error 400: Bad Request
```

## Description

The video https://www.funk.net/channel/doctor-who-1164/journeys-end-ov-515545/doctor-who-staffel-4-ov-2473 plays in a browser (Firefox) without problems, but `youtube-dl` fails to download m3u8 information, MPD & ISM manifests with *HTTP Error 403: Forbidden*.

About 3 weeks ago I didn't experience any problems downloading videos from this site.  Additionally, the problem only seems to occur when the location is **nx-d09**.akamaized.net like
```
[debug] Invoking downloader on u'http://nx-d09.akamaized.net/06be6b49-c771-4f80-9ceb-a658ebc5e751/515545_src_1920x1080_6000.mp4'
ERROR: unable to download video data: HTTP Error 400: Bad Request
```

Other videos, e.g. https://www.funk.net/channel/true-demon-12035/camping-im-wald-nimmt-fuerchterliches-ende-01-1616275/true-demon-alle-folgen-4971 from a different location **funk-01**.akamaized.net can still be downloaded:
```
$ youtube-dl --verbose https://www.funk.net/channel/true-demon-12035/camping-im-wald-nimmt-fuerchterliches-ende-01-1616275/true-demon-alle-folgen-4971

[debug] System config: []
[debug] User config: []
[debug] Custom config: []
[debug] Command-line args: [u'--verbose', u'https://www.funk.net/channel/true-demon-12035/camping-im-wald-nimmt-fuerchterliches-ende-01-1616275/true-demon-alle-folgen-4971']
[debug] Encodings: locale UTF-8, fs UTF-8, out UTF-8, pref UTF-8
[debug] youtube-dl version 2020.03.08
[debug] Python version 2.7.12 (CPython) - Linux-4.15.0-88-generic-x86_64-with-LinuxMint-18.1-serena
[debug] exe versions: ffmpeg 2.8.15, ffprobe 2.8.15, rtmpdump 2.4
[debug] Proxy map: {}
[Funk] 1616275: Downloading JSON metadata
[Nexx] 1616275: Downloading JSON metadata
[Nexx] 1616275: Downloading m3u8 information
[Nexx] 1616275: Downloading MPD manifest
[Nexx] 1616275: Downloading ISM manifest
[debug] Default format spec: bestvideo+bestaudio/best
[debug] Invoking downloader on u'http://funk-01.akamaized.net/098fec4a-c9a8-446e-bcb0-d4ae19f2612c/1616275_src.ism/QualityLevels(5978504)/Manifest(video,format=m3u8-aapl)'
[hlsnative] Downloading m3u8 manifest
[hlsnative] Total fragments: 120
[download] Destination: Camping im Wald nimmt fürchterliches Ende #01-1616275.fazure-hls-6257.mp4
[download]   2.9% of ~567.00MiB at  5.55MiB/s ETA 01:46^C
ERROR: Interrupted by user
```

Thank you for your work.

### PR DESCRIPTION
**Pull request by FliegendeWurst**
### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [adding new extractor tutorial](https://github.com/ytdl-org/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/ytdl-org/youtube-dl#youtube-dl-coding-conventions) sections
- [x] [Searched](https://github.com/ytdl-org/youtube-dl/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests: #24488, but that one will never be merged
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)
- [x] Run the tests (surprisingly enough, many do not)

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)

### What is the purpose of your *pull request*?
- [x] Bug fix

### Description of your *pull request* and other information

See #24309 and #24488

### CODE REVIEW

**Comment by dstftw** [comment]
Nothing changed, still code duplication.

**Comment by FliegendeWurst** [comment]
The duplicate call to _protection_tokens? That's done on two *different* video objects.

**Comment by dstftw** [comment]
No. I've already explained several times, see first reviews.

**Comment by FliegendeWurst** [comment]
So it's this: `x['protectiondata']`. I'll change that.

**Comment by FliegendeWurst** [comment]
@dstftw review again?

**Comment by FliegendeWurst** [comment]
@dstftw I fixed the duplication you mentioned in the [other review](https://github.com/ytdl-org/youtube-dl/pull/24488/files#r400699570)

File: youtube_dl/extractor/nexx.py [hunk 1]
```diff
@@ -220,6 +220,13 @@ def p0(p):
 
         return formats
 
+    def _protection_tokens(self, video):
+        protection_data = video.get('protectiondata')
+        if protection_data:
+            tokenDASH = protection_data.get('tokenDASH')
```
Comment (dstftw) [inline youtube_dl/extractor/nexx.py:226]: Breaks. You are not guaranteed `protection_data` is a dict. **Do not** change previous token extraction algo.

File: youtube_dl/extractor/nexx.py [hunk 2]
```diff
@@ -220,6 +220,13 @@ def p0(p):
...
+            tokenHLS = protection_data.get('tokenHLS')
```
Comment (dstftw) [inline youtube_dl/extractor/nexx.py:227]: DRY.

File: youtube_dl/extractor/nexx.py [hunk 3]
```diff
@@ -247,20 +254,26 @@ def get_cdn_shield_base(shield_type='', static=False):
...
+        tokenDASH, tokenHLS = self._protection_tokens(video)
+        # TODO: is this kind of token storage still in use?
         protection_token = try_get(
             video, lambda x: x['protectiondata']['token'], compat_str)
-        if protection_token:
+        if protection_token and not (tokenHLS or tokenDASH):
             azure_manifest_url += '?hdnts=%s' % protection_token
 
+        azure_manifest_url_DASH = azure_manifest_url + '?hdnts=%s' % (tokenDASH)
+        azure_manifest_url_HLS = azure_manifest_url + '?hdnts=%s' % (tokenHLS)
```
Comment (dstftw) [inline youtube_dl/extractor/nexx.py:265]: DRY.

File: youtube_dl/extractor/nexx.py (hunk 1 above)
Comment (FliegendeWurst) [reply youtube_dl/extractor/nexx.py:226]: I'll use `isinstance` then, thanks. I didn't change the previous token extraction, it simply prefers these new tokens.

File: youtube_dl/extractor/nexx.py (hunk 3 above)
Comment (FliegendeWurst) [reply youtube_dl/extractor/nexx.py:265]: I'll create a function that adds the token to the URL.

File: youtube_dl/extractor/nexx.py (hunk 1 above)
Comment (dstftw) [reply youtube_dl/extractor/nexx.py:226]: Again: no, do not touch token extraction code. Just eliminate duplication.

File: youtube_dl/extractor/nexx.py [hunk 4]
```diff
@@ -69,28 +69,14 @@ class NexxIE(InfoExtractor):
...
+            'alt_title': '1 TAG ohne KLO... wortwörtlich! ?',
+            'description': 'md5:1cf562de6653f76fbc589ecbdf985727',
+            'thumbnail': r're:^https?://.*\.(?:jpg|webp)$',
             'duration': 607,
             'timestamp': 1518614955,
             'upload_date': '20180214',
         },
-    }, {
-        # free cdn from http://www.spiegel.de/video/eifel-zoo-aufregung-um-ausgebrochene-raubtiere-video-99018031.html
-        'url': 'nexx:747:1533779',
```
Comment (FliegendeWurst) [inline youtube_dl/extractor/nexx.py:81]: I moved this test to the spiegel.de extractor (future PR), is that acceptable?

File: youtube_dl/extractor/nexx.py (hunk 1 above)
Comment (FliegendeWurst) [reply youtube_dl/extractor/nexx.py]: I eliminated duplication and had to reorganize the token extraction code. I touched the previous token extraction algo, but did not change the functionality.

File: youtube_dl/extractor/nexx.py (hunk 4 above)
Comment (dstftw) [reply youtube_dl/extractor/nexx.py:81]: No. Spiegel does not match `nexx:`.

File: youtube_dl/extractor/nexx.py (hunk 1 above)
Comment (dstftw) [reply youtube_dl/extractor/nexx.py]: Is this some sort of fat trolling or what?

File: youtube_dl/extractor/nexx.py (hunk 4 above)
Comment (FliegendeWurst) [reply youtube_dl/extractor/nexx.py:81]: Well I obviously changed the url. They appear to no longer use nexx anyway...

File: youtube_dl/extractor/nexx.py (hunk 1 above)
Comment (FliegendeWurst) [reply youtube_dl/extractor/nexx.py]: No, it isn't. Is the new commit not good? It no longer contains any duplication.

Comment (FliegendeWurst) [reply youtube_dl/extractor/nexx.py]: Can you suggest how I am supposed to keep the code DRY **without** at least touching the previous token extractor?

**Comment by FliegendeWurst** [comment]
I have reduced code duplication further and did not change the previous token extraction *algorithm*.

**Comment by FliegendeWurst** [comment]
Closing this PR (doesn't work currently)

**Pull request closed without merging**